### PR TITLE
Queue malice reports

### PR DESCRIPTION
**Pull request by DemiMarie**
This is #98, reopened.

### CODE REVIEW

**Comment by DemiMarie** [comment]
@afck How should we test it?  What about using #101?

**Comment by afck** [comment]
Yes, but possibly #101 will have to be extended to test both cases.

**Comment by DemiMarie** [comment]
We also need to test the case where a node submits a block, even though they are not a validator.

**Comment by varasev** [comment]
I'm going to test this after I make some high priority tasks (describing the contracts in README, Whitepaper, and commenting contracts' code to help auditors understand the contracts easily - I think it may take 1-2 weeks).

**Comment by afck** [comment]
I squashed, rebased and added another commit.

File: ethcore/src/engines/validator_set/safe_contract.rs [hunk 1]
```diff
@@ -295,17 +318,87 @@ impl ValidatorSet for ValidatorSafeContract {
...
+				}
+				if current_block_number > 100 && current_block_number - 100 > block {
+					return false // Report is too old and cannot be used
+				}
+				let (data, decoder) = validator_set::functions::malice_reported_for_block::call(
+					malicious_validator_address, block);
+				let result = client.call_contract(BlockId::Latest, self.contract_address, data)
+					.expect("this is a bug in the genesis block; either the validator set contract is missing, or it is invalid; this is a fatal error in the configuration of Parity, so we cannot recover; qed");
+				decoder.decode(&result[..])
+					.expect("this is a bug in the genesis block; either the validator set contract is missing, or it is invalid; this is a fatal error in the configuration of Parity, so we cannot recover; qed")
```
Comment (vkomenda) [inline ethcore/src/engines/validator_set/safe_contract.rs]: This error message should differ from the one just above.

File: ethcore/src/engines/validator_set/safe_contract.rs [hunk 2]
```diff
@@ -295,17 +318,87 @@ impl ValidatorSet for ValidatorSafeContract {
...
+			debug!(target: "engine", "Checking if report can be removed from cache");
+			let result = {
+				let current_block_number = header.number();
+				if block > current_block_number {
+					return false // Report cannot be used, as it is for a block that isn’t in the current chain
+				}
+				if current_block_number > 100 && current_block_number - 100 > block {
+					return false // Report is too old and cannot be used
+				}
+				let (data, decoder) = validator_set::functions::malice_reported_for_block::call(
```
Comment (vkomenda) [inline ethcore/src/engines/validator_set/safe_contract.rs]: We should include this method in the test validator contract.

File: ethcore/src/engines/validator_set/safe_contract.rs (hunk 1 above)
Comment (afck) [reply ethcore/src/engines/validator_set/safe_contract.rs]: I agree; I'm not even convinced we should panic here: The network could still continue without a malice report queue, and just print a warning. Or is that too dangerous?

File: ethcore/src/engines/validator_set/safe_contract.rs (hunk 2 above)
Comment (vkomenda) [reply ethcore/src/engines/validator_set/safe_contract.rs]: Working on this now.

File: ethcore/src/engines/validator_set/safe_contract.rs (hunk 1 above)
Comment (vkomenda) [reply ethcore/src/engines/validator_set/safe_contract.rs]: A common solution is returning an `Err`. This is the only occurrence if a panic in such a case as far as I can see.

File: ethcore/src/engines/validator_set/safe_contract.rs (hunk 2 above)
Comment (afck) [reply ethcore/src/engines/validator_set/safe_contract.rs]: Ah, sorry, me too! I would have already pushed it, if it hadn't caused the tests to fail in new, original and unexpected ways!
```
test engines::validator_set::contract::tests::reports_validators ... test engines::validator_set::contract::tests::reports_validators has been running for over 60 seconds
```

Comment (afck) [reply ethcore/src/engines/validator_set/safe_contract.rs]: I suspect this is actually a problem with the queue implementation; why else would the test get stuck?
I'll add my contract change to the other PR first, then get back to this issue…

Comment (vkomenda) [reply ethcore/src/engines/validator_set/safe_contract.rs]: Same for me. It gets stuck :(

Comment (afck) [reply ethcore/src/engines/validator_set/safe_contract.rs]: Added to #129 now.

Comment (afck) [reply ethcore/src/engines/validator_set/safe_contract.rs]: And that shows that we really need to get #129 merged before we can continue with this one.

Comment (vkomenda) [reply ethcore/src/engines/validator_set/safe_contract.rs]: And not only `reports_validators`:
```
test engines::validator_set::contract::tests::reports_validators ... test engines::validator_set::contract::tests::reports_validators has been running for over 60 seconds
test ethereum::ethash::tests::can_do_proof_of_work_unordered_verification_fail ... test ethereum::ethash::tests::can_do_proof_of_work_unordered_verification_fail has been running for over 60 seconds
test ethereum::ethash::tests::can_do_seal256_verification_fail ... test ethereum::ethash::tests::can_do_seal256_verification_fail has been running for over 60 seconds
```

**Comment by varasev** [comment]
Can I start to test this with `posdao-test-setup` or there are some fixes waiting to be done?

**Comment by afck** [comment]
Part of this is already in #129.
We still need to figure out how to re-enable and fix the part that's commented out in there; but probably this PR should be closed in favor of #129.

**Comment by varasev** [comment]
> We still need to figure out how to re-enable and fix the part that's commented out in there

Do you mean this? https://github.com/poanetwork/parity-ethereum/blob/74fea45aef174a16d31c0e879216fdcd53c07cef/ethcore/src/engines/validator_set/safe_contract.rs#L344-L362

**Comment by afck** [comment]
Yes, exactly. But I think I already have a solution. Will push it to #129.

**Comment by varasev** [comment]
Done in #129.

**Pull request closed without merging**